titus-server-gateway/src/main/java/com/netflix/titus/gateway/service/v3/internal/LocalCacheQueryProcessor.java

```diff
@@ -209,46 +249,73 @@ private boolean isJobFinishedEvent(JobChangeNotification event) {
-
-                // Job data replicator emits keep alive events periodically if there is nothing in the stream. We have
-                // to filter them out here.
-                if (jobManagerEvent == JobManagerEvent.keepAliveEvent()) {
-                    // Check if staleness is not too high.
-                    if (jobDataReplicator.getStalenessMs() > configuration.getObserveJobsStalenessDisconnectMs()) {
-                        return Mono.error(new StatusRuntimeException(Status.ABORTED.augmentDescription(
-                                "Data staleness in the event stream is too high. Most likely caused by connectivity issue to the downstream server."
-                        )));
-                    }
-                    return Mono.empty();
-                }
-
-                return grpcEvent.map(Flux::just).orElseGet(Flux::empty);
-            });
+            return jobDataReplicator.events()
+                    .subscribeOn(scheduler)
+                    .publishOn(scheduler)
+                    .flatMap(event -> {
+                        JobManagerEvent<?> jobManagerEvent = event.getRight();
+
+                        long now = titusRuntime.getClock().wallTime();
+                        JobSnapshot snapshot = event.getLeft();
+                        Optional<JobChangeNotification> grpcEvent = toObserveJobsEvent(snapshot, jobManagerEvent, now, jobsPredicate, tasksPredicate);
+
+                        // On first event emit full snapshot first
+                        if (first.getAndSet(false)) {
+                            List<JobChangeNotification> snapshotEvents = buildSnapshot(snapshot, now, jobsPredicate, tasksPredicate);
+                            grpcEvent.ifPresent(snapshotEvents::add);
+                            return Flux.fromIterable(snapshotEvents);
+                        }
+
+                        // If we already emitted something we have to first disconnect this stream, and let the client
+                        // subscribe again. Snapshot marker indicates that the underlying GRPC stream was disconnected.
+                        if (jobManagerEvent == JobManagerEvent.snapshotMarker()) {
+                            return Mono.error(new StatusRuntimeException(Status.ABORTED.augmentDescription(
+                                    "Downstream event stream reconnected."
+                            )));
+                        }
+
+                        // Job data replicator emits keep alive events periodically if there is nothing in the stream. We have
+                        // to filter them out here.
+                        if (jobManagerEvent instanceof JobKeepAliveEvent) {
+                            // Check if staleness is not too high.
+                            if (jobDataReplicator.getStalenessMs() > configuration.getObserveJobsStalenessDisconnectMs()) {
+                                return Mono.error(new StatusRuntimeException(Status.ABORTED.augmentDescription(
```

Comment:
Log the observed value of the staleness from the received keep alive event?

Reply:
I added metric to track this event.